### PR TITLE
[DEVHAS-380] Have the Application controller add/remove components from its model

### PR DESCRIPTION
**Pull request by johnmcollier**
### What does this PR do?:
This PR updates the logic that handles the Application's devfile model by moving the code in the Component controller that adds/removes components from it, and moves it to the Application controller. 

With these changes:
- Application reconcile is triggered by a Component creation or deletion
- During each reconcile of an Application, the Application controller retrieves the list of Components that set the Application as its owner, and updates its devfile model accordingly

### Which issue(s)/story(ies) does this PR fixes:
https://issues.redhat.com/browse/DEVHAS-380

### PR acceptance criteria:

- [ ] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

File: controllers/application_controller.go
Comment (maysunfaisal) [inline controllers/application_controller.go:218]: wait why do we do this again

File: controllers/mapper.go
Comment (maysunfaisal) [inline controllers/mapper.go:82]:
```suggestion
		// the obj was not in the namespace or it did not have the required Application.
```
should be?

File: controllers/application_controller.go
Comment (johnmcollier) [reply controllers/application_controller.go:218]: To make comparisons of the added git projects (under projects in the devfile spec) and container image projects (under attributes) easier in the devfile app model. Basically we create a copy of the devworkspace spec, nil out the attributes and projects, add the components to it, and compare with the existing devworkspace spec.

**Review by maysunfaisal — APPROVED** [review]

**Comment by johnmcollier** [comment]
Need to wait for pact tests to be updated before merging

/hold

**Review by yangcao77 — APPROVED** [review]

**Pull request merged**
